Write a program in Python that answers the following question:
How many different 𝑘-mers can be found in the reference genome of the tobacco mosaic virus? Use approximately ten different values for 𝑘, ranging between, say, 4 and 40. What percentage of the maximum number of 𝑘-mers for a given 𝑘 does this represent?
For each 𝑘 that you used, apply 1, 5 and 10% errors (= simple base changes, no indels) along the genome, redo the 𝑘-mer calculation. How did the errors influence the 𝑘-mer count?
For the sequence of the TM-virus, consult the Refseq data base (www.ncbi.nlm.nih.gov/refseq/)

In [1]:
from Bio import Entrez, SeqIO
Entrez.email = "k11850713@students.jku.at" # put your email here

# let's do a search for influenza H1N1 viruses from Texas
handle = Entrez.esearch(db="nucleotide",  # database to search
                        term="tobacco mosaic virus",  # search term
                        retmax=10  # number of results that are returned
                        )
record = Entrez.read(handle)
handle.close()

gi_list = record["IdList"] # list of genbank identifiers found
print (gi_list)

['645985887', '1573711405', '1746316845', '1746316843', '1771120538', '1771094135', '1552774188', '1760664740', '1433342602', '1694998213']


In [2]:
handle = Entrez.efetch(db="nucleotide", id=gi_list, rettype="gb", retmode="text")
records = SeqIO.parse(handle, "genbank")

for record in records:
    print (record.description)
    
handle.close() 

Chain A, Turnip yellow mosaic virus mRNA for the coat protein
Tobacco mosaic virus gene for coat protein, complete cds
Chain R, Cryo-EM structure of TMV with Ca2+ at low pH
Chain R, Cryo-EM structure of TMV in water
WO 2019131426-A/2: A method for direct introduction of a nucleic acid into a plant tissue by electroporation and its product
JP 2019110860-A/2: Production method of transformed plant
Oryza sativa cultivar Carolina Gold Select chromosome 9, whole genome shotgun sequence
Tobacco mosaic virus isolate HN-W5 coat protein (CP) gene, partial cds
Tobacco mosaic virus Jember genomic RNA, contains coat protein
Sunn-hemp mosaic virus 165 kDa protein gene, partial cds; and 30 kDa protein and coat protein genes, complete cds


In [29]:

from Bio import Entrez, SeqIO

Entrez.email = "k11850713@students.jku.at"

handle = Entrez.efetch(db="nucleotide", id="1573711405", rettype="gb", retmode="text")
record = SeqIO.read(handle, "genbank")
handle.close()

# print the record 
print(record)

ID: LC417210.1
Name: LC417210
Description: Tobacco mosaic virus gene for coat protein, complete cds
Number of features: 2
/molecule_type=RNA
/topology=linear
/data_file_division=VRL
/date=16-NOV-2019
/accessions=['LC417210']
/sequence_version=1
/keywords=['']
/source=Tobacco mosaic virus
/organism=Tobacco mosaic virus
/taxonomy=['Viruses', 'Riboviria', 'Virgaviridae', 'Tobamovirus']
/references=[Reference(title='First report of tobacco mosaic virus infecting Vitis coignetiae in Korea', ...), Reference(title='Direct Submission', ...)]
Seq('ATGTCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGCGTGG...TGA', IUPACAmbiguousDNA())


In [30]:
seq = record.seq
seq
myseq = str(seq).replace(' ', '')
myseq

'ATGTCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGCGTGGGCCGACCCAATAGAGTTAATTAATTTATGTACTAATGCCTTAGGAAATCAGTTTCAAACACAACAAGCTCGAACTGTCGTTCAAAGACAATTCACTGAGGTGTGGAAACCTTCACCACAAGTAACTGTTAGGTTCCCTGACAGTGACTTTAAGGTGTACAGGTACAATGCGGTATTAGACCCGCTAGTCACAGCACTGTTGGGTGCATTCGACACTAGAAATAGAATAATAGAAGTTGAAAATCAGGCGAACCCCACGACTGCCGAAACGTTAGATGCTACTCGTAGAGTAGACGACGCAACGGTGGCCATAAGGAGCGCGATAAATAATTTAATAGTAGAATTGATCAGAGGAGCCGGATCTTATAATCGGAGCTCTTTCGAGAGCTCTTCTGGTTTGGTTTGGACCTCTGGTCCTGCAACTTGA'

In [31]:
def build_kmers(sequence, ksize):
    kmers = []
    n_kmers = len(sequence) - ksize + 1

    for i in range(n_kmers):
        kmer = sequence[i:i + ksize]
        kmers.append(kmer)

    return kmers

In [32]:
list_k = [4,6,8,10,12,14,16,18,20,22,30,35,40]

In [33]:
def iterative_k (list_k,seq):
    
    result_list = []
    
    for i in list_k:
        
        result = build_kmers(seq,i)
        result_list.append(result)
    
    return result_list

In [34]:
outcome = iterative_k(list_k,myseq)

In [35]:
def simple(s,n):
    dictionary = {}
    total = 0
    for i in range (len(s)-(n-1)): # (n-1) to get last element
        k = i+n
        if s[i:k] in dictionary:
            dictionary[s[i:k]] += 1
        else:
            dictionary.update({s[i:k]:1})
        total += 1 

    for key, value in dictionary.items():
        dictionary[key] = value/total
    
    dictionary_1 = sorted(dictionary.items(), key=lambda x: x[1], reverse=True)
    dic_list = list(dictionary_1)
    return dic_list

In [36]:
def iterative_k_1 (list_k,seq):
    
    result_list = []
    
    for i in list_k:
        
        result = simple(seq,i)
        result_list.append(result)
    
    return result_list

In [37]:
k_mer_result = iterative_k_1(list_k,myseq)

In [38]:
# Result of 4-kmer (unique subsequence: percentage )
k_mer_result[0]

[('TAGA', 0.018867924528301886),
 ('TAAT', 0.014675052410901468),
 ('AATA', 0.012578616352201259),
 ('TCTT', 0.010482180293501049),
 ('ATCA', 0.010482180293501049),
 ('TCAG', 0.010482180293501049),
 ('AATT', 0.010482180293501049),
 ('GAAA', 0.010482180293501049),
 ('ACAA', 0.010482180293501049),
 ('ACTG', 0.010482180293501049),
 ('ATAA', 0.010482180293501049),
 ('ACAG', 0.008385744234800839),
 ('CAGT', 0.008385744234800839),
 ('AGTA', 0.008385744234800839),
 ('TCAC', 0.008385744234800839),
 ('CACT', 0.008385744234800839),
 ('AGTT', 0.008385744234800839),
 ('GTTC', 0.008385744234800839),
 ('CGAC', 0.008385744234800839),
 ('ATAG', 0.008385744234800839),
 ('AGAG', 0.008385744234800839),
 ('TTAA', 0.008385744234800839),
 ('TTAG', 0.008385744234800839),
 ('AAAT', 0.008385744234800839),
 ('TTCA', 0.008385744234800839),
 ('AGGT', 0.008385744234800839),
 ('GGTG', 0.008385744234800839),
 ('AGAA', 0.008385744234800839),
 ('GAGC', 0.008385744234800839),
 ('TGTC', 0.006289308176100629),
 ('CTTA', 

In [39]:
# total number of subsequence of 4-mers
len(outcome[0])

477

In [40]:
# Numer of unique subsequence of 4-mers
len(k_mer_result[0])

222

In [41]:
# Result of 6-kmer (unique subsequence: percentage )
k_mer_result[1]

[('AATAGA', 0.00631578947368421),
 ('CTACTC', 0.004210526315789474),
 ('TCAGTT', 0.004210526315789474),
 ('TAGAGT', 0.004210526315789474),
 ('TTAATT', 0.004210526315789474),
 ('TAATTT', 0.004210526315789474),
 ('AATTTA', 0.004210526315789474),
 ('AAATCA', 0.004210526315789474),
 ('AATCAG', 0.004210526315789474),
 ('TTCAAA', 0.004210526315789474),
 ('AACTGT', 0.004210526315789474),
 ('AGGTGT', 0.004210526315789474),
 ('ACTGTT', 0.004210526315789474),
 ('ATAGAA', 0.004210526315789474),
 ('TAGAAT', 0.004210526315789474),
 ('AATAAT', 0.004210526315789474),
 ('TAATAG', 0.004210526315789474),
 ('AGTAGA', 0.004210526315789474),
 ('AGGAGC', 0.004210526315789474),
 ('GAGCTC', 0.004210526315789474),
 ('AGCTCT', 0.004210526315789474),
 ('GCTCTT', 0.004210526315789474),
 ('TCTGGT', 0.004210526315789474),
 ('TGGTTT', 0.004210526315789474),
 ('GGTTTG', 0.004210526315789474),
 ('GTTTGG', 0.004210526315789474),
 ('ATGTCT', 0.002105263157894737),
 ('TGTCTT', 0.002105263157894737),
 ('GTCTTA', 0.0021052

In [42]:
# total number of subsequence of 6-mers
len(outcome[1])

475

In [43]:
# Numer of unique subsequence of 6-mers
len(k_mer_result[1])

448

In [44]:
# Result of 8-kmer (unique subsequence: percentage )
k_mer_result[2]

[('GAGCTCTT', 0.004228329809725159),
 ('TGGTTTGG', 0.004228329809725159),
 ('ATGTCTTA', 0.0021141649048625794),
 ('TGTCTTAC', 0.0021141649048625794),
 ('GTCTTACA', 0.0021141649048625794),
 ('TCTTACAG', 0.0021141649048625794),
 ('CTTACAGT', 0.0021141649048625794),
 ('TTACAGTA', 0.0021141649048625794),
 ('TACAGTAT', 0.0021141649048625794),
 ('ACAGTATC', 0.0021141649048625794),
 ('CAGTATCA', 0.0021141649048625794),
 ('AGTATCAC', 0.0021141649048625794),
 ('GTATCACT', 0.0021141649048625794),
 ('TATCACTA', 0.0021141649048625794),
 ('ATCACTAC', 0.0021141649048625794),
 ('TCACTACT', 0.0021141649048625794),
 ('CACTACTC', 0.0021141649048625794),
 ('ACTACTCC', 0.0021141649048625794),
 ('CTACTCCA', 0.0021141649048625794),
 ('TACTCCAT', 0.0021141649048625794),
 ('ACTCCATC', 0.0021141649048625794),
 ('CTCCATCT', 0.0021141649048625794),
 ('TCCATCTC', 0.0021141649048625794),
 ('CCATCTCA', 0.0021141649048625794),
 ('CATCTCAG', 0.0021141649048625794),
 ('ATCTCAGT', 0.0021141649048625794),
 ('TCTCAGTT', 

In [45]:
# total number of subsequence of 8-mers
len(outcome[2])

473

In [46]:
# Numer of unique subsequence of 8-mers
len(k_mer_result[2])

471

In [47]:
# Result of 10-kmer (unique subsequence: percentage )
k_mer_result[3]

[('ATGTCTTACA', 0.0021231422505307855),
 ('TGTCTTACAG', 0.0021231422505307855),
 ('GTCTTACAGT', 0.0021231422505307855),
 ('TCTTACAGTA', 0.0021231422505307855),
 ('CTTACAGTAT', 0.0021231422505307855),
 ('TTACAGTATC', 0.0021231422505307855),
 ('TACAGTATCA', 0.0021231422505307855),
 ('ACAGTATCAC', 0.0021231422505307855),
 ('CAGTATCACT', 0.0021231422505307855),
 ('AGTATCACTA', 0.0021231422505307855),
 ('GTATCACTAC', 0.0021231422505307855),
 ('TATCACTACT', 0.0021231422505307855),
 ('ATCACTACTC', 0.0021231422505307855),
 ('TCACTACTCC', 0.0021231422505307855),
 ('CACTACTCCA', 0.0021231422505307855),
 ('ACTACTCCAT', 0.0021231422505307855),
 ('CTACTCCATC', 0.0021231422505307855),
 ('TACTCCATCT', 0.0021231422505307855),
 ('ACTCCATCTC', 0.0021231422505307855),
 ('CTCCATCTCA', 0.0021231422505307855),
 ('TCCATCTCAG', 0.0021231422505307855),
 ('CCATCTCAGT', 0.0021231422505307855),
 ('CATCTCAGTT', 0.0021231422505307855),
 ('ATCTCAGTTC', 0.0021231422505307855),
 ('TCTCAGTTCG', 0.0021231422505307855),


In [48]:
# total number of subsequence of 10-mers
len(outcome[3])

471

In [49]:
# Numer of unique subsequence of 10-mers
len(k_mer_result[3])

471

In [50]:
# Result of 12-kmer (unique subsequence: percentage )
k_mer_result[4]

[('ATGTCTTACAGT', 0.0021321961620469083),
 ('TGTCTTACAGTA', 0.0021321961620469083),
 ('GTCTTACAGTAT', 0.0021321961620469083),
 ('TCTTACAGTATC', 0.0021321961620469083),
 ('CTTACAGTATCA', 0.0021321961620469083),
 ('TTACAGTATCAC', 0.0021321961620469083),
 ('TACAGTATCACT', 0.0021321961620469083),
 ('ACAGTATCACTA', 0.0021321961620469083),
 ('CAGTATCACTAC', 0.0021321961620469083),
 ('AGTATCACTACT', 0.0021321961620469083),
 ('GTATCACTACTC', 0.0021321961620469083),
 ('TATCACTACTCC', 0.0021321961620469083),
 ('ATCACTACTCCA', 0.0021321961620469083),
 ('TCACTACTCCAT', 0.0021321961620469083),
 ('CACTACTCCATC', 0.0021321961620469083),
 ('ACTACTCCATCT', 0.0021321961620469083),
 ('CTACTCCATCTC', 0.0021321961620469083),
 ('TACTCCATCTCA', 0.0021321961620469083),
 ('ACTCCATCTCAG', 0.0021321961620469083),
 ('CTCCATCTCAGT', 0.0021321961620469083),
 ('TCCATCTCAGTT', 0.0021321961620469083),
 ('CCATCTCAGTTC', 0.0021321961620469083),
 ('CATCTCAGTTCG', 0.0021321961620469083),
 ('ATCTCAGTTCGT', 0.00213219616204

In [51]:
# total number of subsequence of 12-mers
len(outcome[4])

469

In [52]:
# Numer of unique subsequence of 12-mers
len(k_mer_result[4])

469

In [53]:
# Result of 14-kmer (unique subsequence: percentage )
k_mer_result[5]

[('ATGTCTTACAGTAT', 0.0021413276231263384),
 ('TGTCTTACAGTATC', 0.0021413276231263384),
 ('GTCTTACAGTATCA', 0.0021413276231263384),
 ('TCTTACAGTATCAC', 0.0021413276231263384),
 ('CTTACAGTATCACT', 0.0021413276231263384),
 ('TTACAGTATCACTA', 0.0021413276231263384),
 ('TACAGTATCACTAC', 0.0021413276231263384),
 ('ACAGTATCACTACT', 0.0021413276231263384),
 ('CAGTATCACTACTC', 0.0021413276231263384),
 ('AGTATCACTACTCC', 0.0021413276231263384),
 ('GTATCACTACTCCA', 0.0021413276231263384),
 ('TATCACTACTCCAT', 0.0021413276231263384),
 ('ATCACTACTCCATC', 0.0021413276231263384),
 ('TCACTACTCCATCT', 0.0021413276231263384),
 ('CACTACTCCATCTC', 0.0021413276231263384),
 ('ACTACTCCATCTCA', 0.0021413276231263384),
 ('CTACTCCATCTCAG', 0.0021413276231263384),
 ('TACTCCATCTCAGT', 0.0021413276231263384),
 ('ACTCCATCTCAGTT', 0.0021413276231263384),
 ('CTCCATCTCAGTTC', 0.0021413276231263384),
 ('TCCATCTCAGTTCG', 0.0021413276231263384),
 ('CCATCTCAGTTCGT', 0.0021413276231263384),
 ('CATCTCAGTTCGTG', 0.0021413276

In [54]:
# total number of subsequence of 14-mers
len(outcome[5])

467

In [55]:
# Numer of unique subsequence of 14-mers
len(k_mer_result[5])

467

In [56]:
# Result of 16-kmer (unique subsequence: percentage )
k_mer_result[6]

[('ATGTCTTACAGTATCA', 0.002150537634408602),
 ('TGTCTTACAGTATCAC', 0.002150537634408602),
 ('GTCTTACAGTATCACT', 0.002150537634408602),
 ('TCTTACAGTATCACTA', 0.002150537634408602),
 ('CTTACAGTATCACTAC', 0.002150537634408602),
 ('TTACAGTATCACTACT', 0.002150537634408602),
 ('TACAGTATCACTACTC', 0.002150537634408602),
 ('ACAGTATCACTACTCC', 0.002150537634408602),
 ('CAGTATCACTACTCCA', 0.002150537634408602),
 ('AGTATCACTACTCCAT', 0.002150537634408602),
 ('GTATCACTACTCCATC', 0.002150537634408602),
 ('TATCACTACTCCATCT', 0.002150537634408602),
 ('ATCACTACTCCATCTC', 0.002150537634408602),
 ('TCACTACTCCATCTCA', 0.002150537634408602),
 ('CACTACTCCATCTCAG', 0.002150537634408602),
 ('ACTACTCCATCTCAGT', 0.002150537634408602),
 ('CTACTCCATCTCAGTT', 0.002150537634408602),
 ('TACTCCATCTCAGTTC', 0.002150537634408602),
 ('ACTCCATCTCAGTTCG', 0.002150537634408602),
 ('CTCCATCTCAGTTCGT', 0.002150537634408602),
 ('TCCATCTCAGTTCGTG', 0.002150537634408602),
 ('CCATCTCAGTTCGTGT', 0.002150537634408602),
 ('CATCTCA

In [57]:
# total number of subsequence of 16-mers
len(outcome[6])

465

In [58]:
# Numer of unique subsequence of 16-mers
len(k_mer_result[6])

465

In [59]:
# Result of 18-kmer (unique subsequence: percentage )
k_mer_result[7]

[('ATGTCTTACAGTATCACT', 0.0021598272138228943),
 ('TGTCTTACAGTATCACTA', 0.0021598272138228943),
 ('GTCTTACAGTATCACTAC', 0.0021598272138228943),
 ('TCTTACAGTATCACTACT', 0.0021598272138228943),
 ('CTTACAGTATCACTACTC', 0.0021598272138228943),
 ('TTACAGTATCACTACTCC', 0.0021598272138228943),
 ('TACAGTATCACTACTCCA', 0.0021598272138228943),
 ('ACAGTATCACTACTCCAT', 0.0021598272138228943),
 ('CAGTATCACTACTCCATC', 0.0021598272138228943),
 ('AGTATCACTACTCCATCT', 0.0021598272138228943),
 ('GTATCACTACTCCATCTC', 0.0021598272138228943),
 ('TATCACTACTCCATCTCA', 0.0021598272138228943),
 ('ATCACTACTCCATCTCAG', 0.0021598272138228943),
 ('TCACTACTCCATCTCAGT', 0.0021598272138228943),
 ('CACTACTCCATCTCAGTT', 0.0021598272138228943),
 ('ACTACTCCATCTCAGTTC', 0.0021598272138228943),
 ('CTACTCCATCTCAGTTCG', 0.0021598272138228943),
 ('TACTCCATCTCAGTTCGT', 0.0021598272138228943),
 ('ACTCCATCTCAGTTCGTG', 0.0021598272138228943),
 ('CTCCATCTCAGTTCGTGT', 0.0021598272138228943),
 ('TCCATCTCAGTTCGTGTT', 0.00215982721382

In [60]:
# total number of subsequence of 18-mers
len(outcome[7])

463

In [61]:
# Numer of unique subsequence of 18-mers
len(k_mer_result[7])

463

In [62]:
# Result of 20-kmer (unique subsequence: percentage )
k_mer_result[8]

[('ATGTCTTACAGTATCACTAC', 0.0021691973969631237),
 ('TGTCTTACAGTATCACTACT', 0.0021691973969631237),
 ('GTCTTACAGTATCACTACTC', 0.0021691973969631237),
 ('TCTTACAGTATCACTACTCC', 0.0021691973969631237),
 ('CTTACAGTATCACTACTCCA', 0.0021691973969631237),
 ('TTACAGTATCACTACTCCAT', 0.0021691973969631237),
 ('TACAGTATCACTACTCCATC', 0.0021691973969631237),
 ('ACAGTATCACTACTCCATCT', 0.0021691973969631237),
 ('CAGTATCACTACTCCATCTC', 0.0021691973969631237),
 ('AGTATCACTACTCCATCTCA', 0.0021691973969631237),
 ('GTATCACTACTCCATCTCAG', 0.0021691973969631237),
 ('TATCACTACTCCATCTCAGT', 0.0021691973969631237),
 ('ATCACTACTCCATCTCAGTT', 0.0021691973969631237),
 ('TCACTACTCCATCTCAGTTC', 0.0021691973969631237),
 ('CACTACTCCATCTCAGTTCG', 0.0021691973969631237),
 ('ACTACTCCATCTCAGTTCGT', 0.0021691973969631237),
 ('CTACTCCATCTCAGTTCGTG', 0.0021691973969631237),
 ('TACTCCATCTCAGTTCGTGT', 0.0021691973969631237),
 ('ACTCCATCTCAGTTCGTGTT', 0.0021691973969631237),
 ('CTCCATCTCAGTTCGTGTTC', 0.0021691973969631237),


In [63]:
# total number of subsequence of 20-mers
len(outcome[8])

461

In [64]:
# Numer of unique subsequence of 20-mers
len(k_mer_result[8])

461

In [65]:
# Result of 22-kmer (unique subsequence: percentage )
k_mer_result[9]

[('ATGTCTTACAGTATCACTACTC', 0.002178649237472767),
 ('TGTCTTACAGTATCACTACTCC', 0.002178649237472767),
 ('GTCTTACAGTATCACTACTCCA', 0.002178649237472767),
 ('TCTTACAGTATCACTACTCCAT', 0.002178649237472767),
 ('CTTACAGTATCACTACTCCATC', 0.002178649237472767),
 ('TTACAGTATCACTACTCCATCT', 0.002178649237472767),
 ('TACAGTATCACTACTCCATCTC', 0.002178649237472767),
 ('ACAGTATCACTACTCCATCTCA', 0.002178649237472767),
 ('CAGTATCACTACTCCATCTCAG', 0.002178649237472767),
 ('AGTATCACTACTCCATCTCAGT', 0.002178649237472767),
 ('GTATCACTACTCCATCTCAGTT', 0.002178649237472767),
 ('TATCACTACTCCATCTCAGTTC', 0.002178649237472767),
 ('ATCACTACTCCATCTCAGTTCG', 0.002178649237472767),
 ('TCACTACTCCATCTCAGTTCGT', 0.002178649237472767),
 ('CACTACTCCATCTCAGTTCGTG', 0.002178649237472767),
 ('ACTACTCCATCTCAGTTCGTGT', 0.002178649237472767),
 ('CTACTCCATCTCAGTTCGTGTT', 0.002178649237472767),
 ('TACTCCATCTCAGTTCGTGTTC', 0.002178649237472767),
 ('ACTCCATCTCAGTTCGTGTTCT', 0.002178649237472767),
 ('CTCCATCTCAGTTCGTGTTCTT', 0.0

In [66]:
# total number of subsequence of 22-mers
len(outcome[9])

459

In [67]:
# Numer of unique subsequence of 22-mers
len(k_mer_result[9])

459

In [68]:
# Result of 30-kmer (unique subsequence: percentage )
k_mer_result[10]

[('ATGTCTTACAGTATCACTACTCCATCTCAG', 0.0022172949002217295),
 ('TGTCTTACAGTATCACTACTCCATCTCAGT', 0.0022172949002217295),
 ('GTCTTACAGTATCACTACTCCATCTCAGTT', 0.0022172949002217295),
 ('TCTTACAGTATCACTACTCCATCTCAGTTC', 0.0022172949002217295),
 ('CTTACAGTATCACTACTCCATCTCAGTTCG', 0.0022172949002217295),
 ('TTACAGTATCACTACTCCATCTCAGTTCGT', 0.0022172949002217295),
 ('TACAGTATCACTACTCCATCTCAGTTCGTG', 0.0022172949002217295),
 ('ACAGTATCACTACTCCATCTCAGTTCGTGT', 0.0022172949002217295),
 ('CAGTATCACTACTCCATCTCAGTTCGTGTT', 0.0022172949002217295),
 ('AGTATCACTACTCCATCTCAGTTCGTGTTC', 0.0022172949002217295),
 ('GTATCACTACTCCATCTCAGTTCGTGTTCT', 0.0022172949002217295),
 ('TATCACTACTCCATCTCAGTTCGTGTTCTT', 0.0022172949002217295),
 ('ATCACTACTCCATCTCAGTTCGTGTTCTTG', 0.0022172949002217295),
 ('TCACTACTCCATCTCAGTTCGTGTTCTTGT', 0.0022172949002217295),
 ('CACTACTCCATCTCAGTTCGTGTTCTTGTC', 0.0022172949002217295),
 ('ACTACTCCATCTCAGTTCGTGTTCTTGTCA', 0.0022172949002217295),
 ('CTACTCCATCTCAGTTCGTGTTCTTGTCAT', 0.00

In [69]:
# total number of subsequence of 30-mers
len(outcome[10])

451

In [70]:
# Numer of unique subsequence of 30-mers
len(k_mer_result[10])

451

In [71]:
# Result of 35-kmer (unique subsequence: percentage )
k_mer_result[11]

[('ATGTCTTACAGTATCACTACTCCATCTCAGTTCGT', 0.002242152466367713),
 ('TGTCTTACAGTATCACTACTCCATCTCAGTTCGTG', 0.002242152466367713),
 ('GTCTTACAGTATCACTACTCCATCTCAGTTCGTGT', 0.002242152466367713),
 ('TCTTACAGTATCACTACTCCATCTCAGTTCGTGTT', 0.002242152466367713),
 ('CTTACAGTATCACTACTCCATCTCAGTTCGTGTTC', 0.002242152466367713),
 ('TTACAGTATCACTACTCCATCTCAGTTCGTGTTCT', 0.002242152466367713),
 ('TACAGTATCACTACTCCATCTCAGTTCGTGTTCTT', 0.002242152466367713),
 ('ACAGTATCACTACTCCATCTCAGTTCGTGTTCTTG', 0.002242152466367713),
 ('CAGTATCACTACTCCATCTCAGTTCGTGTTCTTGT', 0.002242152466367713),
 ('AGTATCACTACTCCATCTCAGTTCGTGTTCTTGTC', 0.002242152466367713),
 ('GTATCACTACTCCATCTCAGTTCGTGTTCTTGTCA', 0.002242152466367713),
 ('TATCACTACTCCATCTCAGTTCGTGTTCTTGTCAT', 0.002242152466367713),
 ('ATCACTACTCCATCTCAGTTCGTGTTCTTGTCATC', 0.002242152466367713),
 ('TCACTACTCCATCTCAGTTCGTGTTCTTGTCATCA', 0.002242152466367713),
 ('CACTACTCCATCTCAGTTCGTGTTCTTGTCATCAG', 0.002242152466367713),
 ('ACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGC',

In [72]:
# total number of subsequence of 35-mers
len(outcome[11])

446

In [73]:
# Numer of unique subsequence of 35-mers
len(k_mer_result[11])

446

In [74]:
# Result of 40-kmer (unique subsequence: percentage )
k_mer_result[12]

[('ATGTCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCT', 0.0022675736961451248),
 ('TGTCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTT', 0.0022675736961451248),
 ('GTCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTG', 0.0022675736961451248),
 ('TCTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGT', 0.0022675736961451248),
 ('CTTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTC', 0.0022675736961451248),
 ('TTACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCA', 0.0022675736961451248),
 ('TACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCAT', 0.0022675736961451248),
 ('ACAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATC', 0.0022675736961451248),
 ('CAGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCA', 0.0022675736961451248),
 ('AGTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAG', 0.0022675736961451248),
 ('GTATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGC', 0.0022675736961451248),
 ('TATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGCG', 0.0022675736961451248),
 ('ATCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGCGT', 0.0022675736961451248),
 ('TCACTACTCCATCTCAGTTCGTGTTCTTGTCATCAGCGTG', 0.0022675736961451248),
 ('CACTACTCCATCTCAGT

In [75]:
# total number of subsequence of 40-mers
len(outcome[12])

441

In [76]:
# Numer of unique subsequence of 40-mers
len(k_mer_result[12])

441

In [80]:
# To compute the percentage of unique k-mers compared to total number of k-mer subsequences for each k

unique_percent = []

for a in range (0,12):
    uni_percent = len(k_mer_result[a]) /len(outcome[a])
    unique_percent.append(uni_percent)

In [82]:
# 4-kmer has around 47% unique k-mer subsequence. After 10-mers, all k-mers has 100% 
unique_percent

[0.46540880503144655,
 0.9431578947368421,
 0.9957716701902748,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0]

In [95]:
seq_list = list(myseq)

In [96]:
import random
def mutate_thre(string, mutation, threshold):
    char =  random.randint(0, len(string)-1) 
    return ''.join([mutation[char] if random.random() < threshold and char in mutation else char for char in string])

In [97]:
# error 1% 

selection = ['A','T','C','G']

index = random.choice(selection)
index_a = random.choice(selection) 
myseq_01 = mutate_thre(seq_list,{index:index_a}, 0.01)

In [103]:
outcome_01 = iterative_k(list_k,myseq_01)
k_mer_result_01 = iterative_k_1(list_k,myseq_01)

In [104]:
unique_percent_01 = []

for a in range (0,12):
    uni_percent_01 = len(k_mer_result_01[a]) /len(outcome_01[a])
    unique_percent_01.append(uni_percent_01)

In [105]:
unique_percent_01

[0.46540880503144655,
 0.9452631578947368,
 0.9957716701902748,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0]

In [143]:
res_01 = "\n".join("{}    {}".format(len(x), len(y)) for x, y in zip(outcome_01,k_mer_result_01))

In [144]:
# Error 1%
print ("Total  Unique")
print(res_01)

Total  Unique
477    222
475    449
473    471
471    471
469    469
467    467
465    465
463    463
461    461
459    459
451    451
446    446
441    441


In [107]:
# error 5%
selection = ['A','T','C','G']

index = random.choice(selection)
index_a = random.choice(selection) 
myseq_05 = mutate_thre(seq_list,{index:index_a}, 0.05)

In [108]:
outcome_05 = iterative_k(list_k,myseq_05)
k_mer_result_05 = iterative_k_1(list_k,myseq_05)

In [109]:
unique_percent_05 = []

for a in range (0,12):
    uni_percent_05 = len(k_mer_result_05[a]) /len(outcome_05[a])
    unique_percent_05.append(uni_percent_05)

In [110]:
unique_percent_05

[0.46540880503144655,
 0.9431578947368421,
 0.9957716701902748,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0]

In [145]:
res_05 = "\n".join("{}    {}".format(len(x), len(y)) for x, y in zip(outcome_05,k_mer_result_05))

In [146]:
# Error 5%
print ("Total  Unique")
print(res_05)

Total  Unique
477    222
475    448
473    471
471    471
469    469
467    467
465    465
463    463
461    461
459    459
451    451
446    446
441    441


In [150]:
# error 10%
selection = ['A','T','C','G']

index = random.choice(selection)
index_a = random.choice(selection) 
myseq_10 = mutate_thre(seq_list,{index:index_a}, 0.1)

In [151]:
outcome_10 = iterative_k(list_k,myseq_10)
k_mer_result_10 = iterative_k_1(list_k,myseq_10)

In [152]:
unique_percent_10 = []

for a in range (0,12):
    uni_percent_10 = len(k_mer_result_10[a]) /len(outcome_10[a])
    unique_percent_10.append(uni_percent_10)

In [153]:
unique_percent_10

[0.4549266247379455,
 0.9389473684210526,
 0.9978858350951374,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0,
 1.0]

In [154]:
res_10 = "\n".join("{}    {}".format(len(x), len(y)) for x, y in zip(outcome_10,k_mer_result_10))

In [149]:
# Error 10 % 
print ("Total  Unique")
print(res_10)

Total  Unique
477    221
475    450
473    471
471    471
469    469
467    467
465    465
463    463
461    461
459    459
451    451
446    446
441    441


In [163]:
res_ref = "\n".join("{}    {}".format(len(x), len(y)) for x, y in zip(outcome,k_mer_result))

In [164]:
# Reference sequence 
print ("Total  Unique")
print(res_ref)

Total  Unique
477    222
475    448
473    471
471    471
469    469
467    467
465    465
463    463
461    461
459    459
451    451
446    446
441    441


Based on the three results from the three different errors, we have found no significant difference in both total and unique number of k-mer count for each tested k number. To be more detailed, with error of 1%, the unique subsequence number of 6-mer has been changed to 449. As well, again with 6-kmer, the unique subsequence number has been changed to 450 with 10% error modified. The unique subsequence number with no error for 6-mer is 448. In addition, the unique number of 4-mer has been changed to 221 compared to 222 with the reference sequence. 
Overall, there is no significant influence identified with the three different tested error percentage for k-mers 